Staged build

mtzguido opened the pull request:
This PR removes all generated OCaml files and stages the build. After `make`, it installs pulse into an `out/` directory, mimicking the binary package exactly. Running `make` will check the plugin sources (src/), extract them and build them, and then continue checking the library.

The library is also reorganized into `core`, `common`, and `pulse`: `common` contains interface modules like FractionalPermission and Pulse.Lib.Core.fsti; `pulse` contains user-facing Pulse files, like Reference and Array, and `core` contains the model. On a normal `make` run, the default rule will not check or build `core`, to save some time. Running `make all` will check these files, possibly in parallel with everything else (including the building of the plugin and the `pulse` library).

tahina-pro commented on README.md:
```diff
@@ -15,10 +15,10 @@ the following cases:
+  * the theoretical foundations of Pulse, in `core/` (namespace `PulseCore`)
+  * the interface powering the Pulse DSL, `common/Pulse.Main.fsti`
+  * the Pulse standard library, in `pulse/` (namespace `Pulse.Lib`), with
     some subdirectories for typeclasses (in `class`, namespace
```
Thanks Guido for your proposal!

With this PR, do lines 7-13 still hold?
> This repository has been designed to closely follow the Filesystem
Hierarchy Standard (FHS), so that it can be used the same way in all
the following cases:
> * direct use of the Pulse repository clone
> * installation into a FHS-like hierarchy (e.g. `/usr/local`)
> * installation as an _opam_ package, using the _opam_ package manager for
  OCaml

From what I understand, for this to hold, only the Pulse standard library would be installed via `make install`, and elaborated code produced by the Pulse elaborator should not depend on `lib/common`. Is my understanding correct? If so, this should be documented somewhere.

If that does not hold, how should a library developer depending on Pulse configure the location of the Pulse library in their Makefiles and how should they detect whether their users installed Pulse with opam or `make install` or clone it from the GitHub repository?

mtzguido replied:
`lib/common` is also installed (it must be since Pulse.Lib.Core.fsti is there). What we skip is installing PulseCore implementation files (though the checked files are installed... this is the case now too, and wasteful, so I'll remove it in a later PR).

All a user has to do is `--include /xyz/lib/pulse`, where `/xyz` is the root (PREFIX) of the installation. If they just build the repo without installing it, this is `out/lib/pulse``. If they make install, it is `/usr/local/lib/pulse`. And for OPAM it's `/home/user/.opam/sw/lib/pulse`.

So they do not really have to distinguish these cases at all, just point to the right `lib/pulse` directory, like until now? With the change that now there is an extra `out/` component for the local case.

I changed the README slightly. One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.

tahina-pro replied:
> `lib/common` is also installed [...]
> All a user has to do is `--include /xyz/lib/pulse`, where `/xyz` is the root (PREFIX) of the installation. If they just build the repo without installing it, this is `` out/lib/pulse``. If they make install, it is ``/usr/local/lib/pulse`. And for OPAM it's `/home/user/.opam/sw/lib/pulse`.

1. does this mean that this single `--include` is enough for users, and `fstar.include` files from there take care of `include`-ing `lib/common`?
2. where is `lib/common` installed? I believe it cannot be installed as `/home/user/.opam/sw/lib/common` (and similarly for `make install`) and it has to go under `lib/pulse` instead

> With the change that now there is an extra out/ component for the local case.

Thanks Guido for clarifying this case. Then, I think it would be simpler to remove that now confusing "FHS" introductory paragraph from README.md: the structure of the repository now no longer matters since people using the repository should use the `out` subdirectory instead of the root directory of the repository.

>  One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.

Don't we also have `--load_cmxs pulse` ?

mtzguido replied:
> 1. does this mean that this single `--include` is enough for users, and `fstar.include` files from there take care of `include`-ing `lib/common`?

Sort of, but indeed the single include is enough. The file `lib/pulse/fstar.include` contains only `lib`, but everything is there in `lib/pulse/lib`. The directories `common` and `pulse` are flattened in the package, users don't ever see them, it's a convention of the source repo, not the package.

> 2. where is `lib/common` installed? I believe it cannot be installed as `/home/user/.opam/sw/lib/common` (and similarly for `make install`) and it has to go under `lib/pulse` instead

It all goes into `lib/pulse/lib`. I suppose a better name could be `lib/pulse/fstar`, since this directory has all F*-specific files (fst/fsti/checked). The parent `lib/pulse` has the built OCaml files, including `pulse.cmxs`.
 
> > With the change that now there is an extra out/ component for the local case.
> 
> Thanks Guido for clarifying this case. Then, I think it would be simpler to remove that now confusing "FHS" introductory paragraph from README.md: the structure of the repository now no longer matters since people using the repository should use the `out` subdirectory instead of the root directory of the repository.

Sounds good!

> 
> > One thing I removed is `share/pulse/Makefile.include`, do we still want this? Given that the only flag users need is a single `--include`, I don't see why we want Pulse-specific Makefiles.
> 
> Don't we also have `--load_cmxs pulse` ?

This is not needed since `#lang-pulse` will trigger F* to find the `pulse.cmxs` plugin in the include path and automatically load it.

gebner commented:
@mtzguido What is your plan for the `FSTAR_HOME` environment variables in `src/extraction/Extraction.fst.config.json`, etc.?

mtzguido commented:
> @mtzguido What is your plan for the `FSTAR_HOME` environment variables in `src/extraction/Extraction.fst.config.json`, etc.?

These remain as-is, Pulse still requires you to define an `FSTAR_HOME` so we can find the sources and checked files for the compiler. The Makefile will also error out quick if this is unset.

mtzguido commented:
I think this is ready. @gebner @tahina-pro , any objections to merging now?

gebner commented:
Please restore the extraction tests that were in `share/pulse/examples`.

> the default rule will not check or build core, to save some time.

I'm not sure if this really makes a difference...  The main build time is the checker now (and it would be great to have lax checking there by default).  Core doesn't really matter as it can be built in parallel with the library.

For some probably related reason the pulse2rust test also builds core from scratch now.

mtzguido commented:
> Please restore the extraction tests that were in `share/pulse/examples`.

Done, sorry I missed that. They are now also checked against an expected output, if that becomes annoying we can just remove those files and add `EXTRACT_C += Module.fst` where relevant. I moved them to `tests/` too.

> 
> > the default rule will not check or build core, to save some time.
> 
> I'm not sure if this really makes a difference... The main build time is the checker now (and it would be great to have lax checking there by default). Core doesn't really matter as it can be built in parallel with the library.

It's not much, agree... maybe forgetting to check core is more of a problem than the slight time saved. So I added it back to the default.

As for lax-checking the sources, that would be nice, but I think the clean way of doing it would need some F* support, since currently we will just get normal-looking checked files. Probably the nicest would be decoupling the generation of the .checked with the module's verification.

> For some probably related reason the pulse2rust test also builds core from scratch now.

Thanks this was an --include that was too big and so lib/core/ was being read by the dependency analysis.

gebner commented:
Great!  I'm happy with merging then.

mtzguido commented:
Going for it.. Tahina let me know if any of your stuff is affected by this and I'll be happy to take a look.
